iocore/eventsystem/IOBuffer.cc

```diff
@@ -271,3 +272,74 @@ IOBufferReader::memcpy(const void *ap, int64_t len, int64_t offset)
 
   return p;
 }
+
+void *
+ats_malloc_pool(size_t size)
+{
+  void *ptr = NULL;
+  size_t total_required_size = size + sizeof(uint64_t);
+
+  if (likely(size > 0)) {
+    if (unlikely(total_required_size > DEFAULT_MAX_BUFFER_SIZE)) {
+      // we need to fall back to ats_malloc for large allocations
+      ptr = ats_malloc(total_required_size);
+    } else {
+      int64_t iobuffer_index = iobuffer_size_to_index(total_required_size);
+      ink_release_assert(iobuffer_index >= 0);
+
+      ptr = ioBufAllocator[iobuffer_index].alloc_void();
+      Debug("allocator_pool",
+            "Allocating a block from iobuf index %ld (block size %ld), requested_size = %ld, alloc size = %ld at location %p",
+            iobuffer_index, index_to_buffer_size(iobuffer_index), size, total_required_size, ptr);
+    }
+  } else {
+    return NULL;
+  }
```

Comment:
In ``ats_realloc_pool`` you do an early return on NULL, I think you should do the same here.